File: dp_labs_and_solutions/dp_labs_and_solutions/dp_java_labs_and_solutions/abstract_factory/lab/Garden.java
// Garden.java

public class Garden
{
	public Garden(int aLettuceType, int aTomatoType)
	{
		setLettuceType(aLettuceType);
		setTomatoType(aTomatoType);
	}

	public void setLettuceType(int aLettuceType)
	{
		if (aLettuceType != 1 && aLettuceType != 2)
		{
			throw new DpException("void setLettuceType(int aLettuceType) -- A Garden must have a valid Lettuce type.");
		}

		myLettuceType = aLettuceType;
	}

	public void setTomatoType(int aTomatoType)
	{
		if (aTomatoType != 1 && aTomatoType != 2)
		{
			throw new DpException("void setTomatoType(int aTomatoType) -- A Garden must have a valid Tomato type.");
		}

		myTomatoType = aTomatoType;
	}

	public Lettuce makeLettuce()
	{
		if (myLettuceType == 1)
		{
			return new ConventionalLettuce(1, 2);
		}
		else if (myLettuceType == 2)
		{
			return new OrganicLettuce(1, 2);
		}
		else
		{
			throw new DpException("Lettuce makeLettuce() -- Oops, should never reach this point.");
		}
	}

	public Tomato makeTomato()
	{
		if (myTomatoType == 1)
		{
			return new ConventionalTomato(1, 2);
		}
		else if (myTomatoType == 2)
		{
			return new OrganicTomato(1, 2);
		}
		else
		{
			throw new DpException("Tomato makeTomato() -- Oops, should never reach this point.");
		}
	}

	private int myLettuceType;
	private int myTomatoType;
}

File: dp_labs_and_solutions/original/dp_labs_and_solutions/dp_cpp_labs_and_solutions/factory_method/lab/garden.cpp
// garden.cpp

#include "conventional_tomato.h"
#include "dp_exception.h"
#include "organic_tomato.h"
#include "garden.h"
#include "tomato.h"

garden::garden(int a_tomato_type)
{
	set_tomato_type(a_tomato_type);
}

void garden::set_tomato_type(int a_tomato_type)
{
	if (a_tomato_type != 1 && a_tomato_type != 2)
	{
		throw dp_exception("void garden::set_tomato_type(int a_tomato_type) -- A garden must have a valid tomato type.");
	}

	my_tomato_type = a_tomato_type;
}

tomato * garden::make_tomato()
{
	if (my_tomato_type == 1)
	{
		return new conventional_tomato(1, 2);
	}
	else if (my_tomato_type == 2)
	{
		return new organic_tomato(1, 2);
	}
	else
	{
		throw dp_exception("tomato * garden::make_tomato() -- Oops, should never reach this point.");
	}
}
